File: ios/GradeThread/Marketplaces/Reconciliation/ReconciliationService.swift
import Foundation
import Supabase

/// supabase-swift wrapper for the orphan-listing reconciliation flow.
/// Three actions per orphan: Create (insert inventory_items + flip
/// matched), Link (point at an existing item + flip matched), Ignore
/// (flip ignored).
@MainActor
public final class ReconciliationService {

    private let supabase: SupabaseClient

    /// Cached writer for the `listed_at` timestamp. Allocating an
    /// `ISO8601DateFormatter` is expensive and an insert can happen per orphan
    /// in a batch, so we reuse one instead of building a fresh formatter per
    /// insert (US-1226, mirrors the US-1007 fix in `EbaySyncService`).
    /// `.withFractionalSeconds` so iOS-written timestamps match server writes.
    private static let isoWriter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    nonisolated public init(supabase: SupabaseClient = SupabaseShared.client) {
        self.supabase = supabase
    }

    // MARK: - Read

    /// Lightweight count of unmatched eBay listings — used by the shell-level
    /// Reconcile affordance (US-749) so the badge can surface on any tab without
    /// loading the full orphan list. Mirrors `fetchOrphans`'s filter.
    public func countOrphans(userId: String) async throws -> Int {
        // US-1649: HEAD + exact count — the server returns just the count, not
        // every unmatched row (the old .select("id") downloaded the whole set
        // only to take rows.count, which scaled badly with orphan volume).
        let response = try await supabase
            .from("flipdesk_ebay_listings")
            .select("id", head: true, count: .exact)
            .eq("user_id", value: userId)
            .eq("match_status", value: "unmatched")
            .execute()
        return response.count ?? 0
    }

    /// Fetches every unmatched eBay listing for the user, newest first.
    public func fetchOrphans(userId: String) async throws -> [OrphanEbayListing] {
        try await supabase
            .from("flipdesk_ebay_listings")
            .select("id, ebay_item_id, custom_label, title, current_price, available_quantity, listing_url, listing_format, imported_at")
            .eq("user_id", value: userId)
            .eq("match_status", value: "unmatched")
            .order("imported_at", ascending: false)
            .execute()
            .value
    }

    // MARK: - Mutations

    /// Inserts a new `inventory_items` row from the orphan's fields and
    /// flips the eBay row to matched + linked. The caller passes the
    /// user-edited title / SKU / target price so Create-sheet edits flow
    /// through; pass nil to fall back to the orphan defaults.
    public func createItem(
        from orphan: OrphanEbayListing,
        userId: String,
        title: String? = nil,
        sku: String? = nil,
        targetPrice: Double? = nil
    ) async -> ReconciliationOutcome {
        struct ItemInsert: Encodable {
            // US-1648: a client-generated id gives the row a stable identity, so
            // the write below is an idempotent UPSERT — a retry after a lost
            // response updates the SAME row instead of minting a duplicate
            // inventory item (matches the photo-first draft convention).
            let id: String
            let user_id: String
            let title: String
            let sku: String?
            let target_price: Double?
            let status: String
            let item_category: String?
        }

        let finalTitle = (title?.trimmingCharacters(in: .whitespaces).nonEmpty
            ?? orphan.suggestedTitle)
        let finalSku = sku?.trimmingCharacters(in: .whitespaces).nonEmpty
            ?? orphan.customLabel?.trimmingCharacters(in: .whitespaces).nonEmpty
        let finalTarget = targetPrice ?? orphan.currentPrice
        // Lowercased to match Postgres uuid text form (the app's convention).
        let newId = UUID().uuidString.lowercased()

        do {
            let payload = ItemInsert(
                id: newId,
                user_id: userId,
                title: finalTitle,
                sku: finalSku,
                target_price: finalTarget,
                // eBay says these are listed — start them in 'listed'
                // status so they show up correctly in the inventory list.
                status: "listed",
                item_category: nil
            )
            try await supabase
                .from("inventory_items")
                .upsert(payload)
                .execute()
            // US-751: mirror the live eBay listing into `listings` (best-effort)
            // so the new item's canvas immediately shows where it's listed — the
            // web reconciliation does this, iOS previously didn't, so a created
            // item showed 0 listings. A failure here must NOT strand the created
            // item, so it's swallowed; the next eBay sync also reconciles it.
            try? await upsertEbayListingRow(forItemId: newId, from: orphan)
            try await markMatched(orphanId: orphan.id, matchedItemId: newId)
            return ReconciliationOutcome(orphanId: orphan.id, kind: .created(itemId: newId))
        } catch {
            return ReconciliationOutcome(orphanId: orphan.id, kind: .failed(message: error.localizedDescription))
        }
    }

    /// Links an orphan to an existing inventory_items row. Mirrors the live eBay
    /// listing into `listings` (US-751) so the linked item's canvas shows it,
    /// then flips match_status + matched_item_id. A listing-mirror failure
    /// surfaces as `.failed` (the upsert is select-then-insert/update, so a retry
    /// can't duplicate the row).
    ///
    /// US-1509: refuses to attach a SECOND active eBay listing to an item that
    /// already has one under a different eBay item id — that's how duplicate
    /// active rows are born, which then feed the wrong-row bulk-action class.
    /// Re-linking the SAME listing id stays allowed (the mirror upsert is
    /// idempotent). On success the item's status advances to "listed",
    /// consistent with `createItem` (eBay says it's live).
    public func link(
        orphan: OrphanEbayListing,
        toExistingItemId itemId: String
    ) async -> ReconciliationOutcome {
        do {
            struct ActiveRow: Decodable { let platform_listing_id: String? }
            let active: [ActiveRow] = try await supabase
                .from("listings")
                .select("platform_listing_id")
                .eq("inventory_item_id", value: itemId)
                .eq("platform", value: "ebay")
                .eq("is_active", value: true)
                .execute()
                .value
            if active.contains(where: { $0.platform_listing_id != orphan.ebayItemId }) {
                return ReconciliationOutcome(
                    orphanId: orphan.id,
                    kind: .failed(message: "That item already has an active eBay listing. Link this listing to a different item, or end the item\u{2019}s existing listing first.")
                )
            }
            try await upsertEbayListingRow(forItemId: itemId, from: orphan)
            try await markMatched(orphanId: orphan.id, matchedItemId: itemId)
            // Best-effort like the US-751 mirror: a status-write blip must not
            // strand the completed link — the next eBay sync reconciles it too.
            try? await markItemListed(itemId: itemId)
            return ReconciliationOutcome(orphanId: orphan.id, kind: .linked(itemId: itemId))
        } catch {
            return ReconciliationOutcome(orphanId: orphan.id, kind: .failed(message: error.localizedDescription))
        }
    }

    /// Marks an orphan as ignored — won't appear in the queue again
    /// until a re-sync overwrites it.
    public func ignore(orphan: OrphanEbayListing) async -> ReconciliationOutcome {
        struct Update: Encodable { let match_status: String }
        do {
            // US-1226: scope the update to the signed-in user as well as the row
            // id, for defense-in-depth parity with the US-268 explicit-scoping
            // convention (a stale/forged id alone never flips another tenant's row).
            let userId = try await currentUserId()
            try await supabase
                .from("flipdesk_ebay_listings")
                .update(Update(match_status: "ignored"))
                .eq("id", value: orphan.id)
                .eq("user_id", value: userId)
                .execute()
            return ReconciliationOutcome(orphanId: orphan.id, kind: .ignored)
        } catch {
            return ReconciliationOutcome(orphanId: orphan.id, kind: .failed(message: error.localizedDescription))
        }
    }

    /// Maximum number of orphans to create concurrently in `createAll`. Mirrors
    /// `DraftsLibraryStore.bulkSaveConcurrency` (US-1166): each create is an
    /// independent multi-row write (inventory_items insert + listing mirror +
    /// match flip) with no shared mutable state, so a small bounded fan-out turns
    /// the old multi-second-per-item serial march into overlapping round-trips
    /// without flooding the connection pool.
    static let bulkCreateConcurrency = 4

    /// Runs `createItem` for every orphan with bounded concurrency (US-1240).
    /// Per-row failures are collected (one bad row doesn't stop the batch), and
    /// `progress` is invoked on the main actor with `(done, total)` after each
    /// chunk resolves so the UI can surface incremental progress instead of a
    /// single opaque spinner.
    public func createAll(
        _ orphans: [OrphanEbayListing],
        userId: String,
        progress: (@MainActor (Int, Int) -> Void)? = nil
    ) async -> ReconciliationBulkResult {
        let total = orphans.count
        guard total > 0 else { return ReconciliationBulkResult(succeeded: 0, failures: []) }

        var succeeded = 0
        var failures: [(String, String)] = []
        var done = 0
        progress?(0, total)

        // US-1240: create in bounded-concurrency chunks instead of awaiting each
        // createItem serially. Each create is independent, so the chunked task
        // group is safe; results fold back here on the @MainActor service.
        for chunk in orphans.chunked(into: Self.bulkCreateConcurrency) {
            let results = await withTaskGroup(
                of: (orphanId: String, ok: Bool, message: String?).self
            ) { group -> [(orphanId: String, ok: Bool, message: String?)] in
                for orphan in chunk {
                    group.addTask {
                        let outcome = await self.createItem(from: orphan, userId: userId)
                        switch outcome.kind {
                        case .created:
                            return (orphan.id, true, nil)
                        case .failed(let message):
                            return (orphan.id, false, message)
                        case .linked, .ignored:
                            // createItem only ever returns .created/.failed; guard.
                            return (orphan.id, false, "Unexpected outcome.")
                        }
                    }
                }
                var acc: [(orphanId: String, ok: Bool, message: String?)] = []
                for await r in group { acc.append(r) }
                return acc
            }
            for r in results {
                done += 1
                if r.ok { succeeded += 1 } else { failures.append((r.orphanId, r.message ?? "Create failed.")) }
            }
            progress?(done, total)
        }
        return ReconciliationBulkResult(succeeded: succeeded, failures: failures)
    }

    // MARK: - eBay listing mirror (US-751)

    /// The `listings`-row fields derived from an orphan eBay listing. Extracted as
    /// a pure mapping so the create/link mirror contract is unit-testable without
    /// Supabase. Mirrors the web's `upsertEbayListingRowForItem` (`listing_price`
    /// defaults to 0 when the orphan has no price; title/url pass through).
    public struct EbayListingMirrorFields: Equatable {
        public let platformListingId: String
        public let listingURL: String?
        public let listingPrice: Double
        public let listingTitle: String?
    }

    public static func ebayListingMirrorFields(
        from orphan: OrphanEbayListing
    ) -> EbayListingMirrorFields {
        EbayListingMirrorFields(
            platformListingId: orphan.ebayItemId,
            listingURL: orphan.listingURL,
            listingPrice: orphan.currentPrice ?? 0,
            listingTitle: orphan.title
        )
    }

    // MARK: - Private

    /// Mirrors a live eBay listing into the `listings` table for `itemId`, so the
    /// item canvas (which reads `listings`/`LocalListing`) shows where the item is
    /// listed after a reconcile. Idempotent: looks up an existing eBay listing
    /// for this item+ebay id and updates it, else inserts a fresh one — matching
    /// the web reconciliation (`upsertEbayListingRowForItem`). RLS scopes
    /// `listings` through the parent `inventory_items` row, which the caller has
    /// just created or owns, so no explicit user_id is needed (the table has none).
    private func upsertEbayListingRow(
        forItemId itemId: String,
        from orphan: OrphanEbayListing
    ) async throws {
        struct ExistingId: Decodable { let id: String }
        // Common, server-authoritative fields shared by insert + update. `listed_at`
        // is set ONLY on insert so an update never clobbers the original list date.
        struct Update: Encodable {
            let platform = "ebay"
            // US-1077: matching an imported live eBay listing → eBay-originated.
            let listing_origin = "ebay"
            let listing_url: String?
            let listing_price: Double
            let listing_title: String?
            let listing_status = "active"
            let is_active = true
        }
        struct Insert: Encodable {
            let inventory_item_id: String
            let platform = "ebay"
            let listing_origin = "ebay"
            let platform_listing_id: String
            let listing_url: String?
            let listing_price: Double
            let listing_title: String?
            let listing_status = "active"
            let is_active = true
            let listed_at: String
        }

        let fields = Self.ebayListingMirrorFields(from: orphan)
        let existing: [ExistingId] = try await supabase
            .from("listings")
            .select("id")
            .eq("inventory_item_id", value: itemId)
            .eq("platform", value: "ebay")
            .eq("platform_listing_id", value: fields.platformListingId)
            .limit(1)
            .execute()
            .value

        if let first = existing.first {
            try await supabase
                .from("listings")
                .update(Update(
                    listing_url: fields.listingURL,
                    listing_price: fields.listingPrice,
                    listing_title: fields.listingTitle
                ))
                .eq("id", value: first.id)
                .execute()
        } else {
            let listedAt = Self.isoWriter.string(from: Date())
            try await supabase
                .from("listings")
                .insert(Insert(
                    inventory_item_id: itemId,
                    platform_listing_id: fields.platformListingId,
                    listing_url: fields.listingURL,
                    listing_price: fields.listingPrice,
                    listing_title: fields.listingTitle,
                    listed_at: listedAt
                ))
                .execute()
        }
    }

    /// US-1509: advance a just-linked item to "listed", matching the status
    /// `createItem` starts created items in — eBay says the listing is live, so
    /// the item must not linger in a pre-list pipeline stage (where the canvas
    /// offers a publish CTA that would double-list it). RLS scopes the row to
    /// the signed-in user.
    private func markItemListed(itemId: String) async throws {
        struct Update: Encodable { let status = "listed" }
        try await supabase
            .from("inventory_items")
            .update(Update())
            .eq("id", value: itemId)
            .execute()
    }

    private func markMatched(orphanId: String, matchedItemId: String) async throws {
        struct Update: Encodable {
            let match_status: String
            let matched_item_id: String
        }
        // US-1226: scope to the signed-in user in addition to the row id, for
        // defense-in-depth parity with the US-268 explicit-scoping convention.
        let userId = try await currentUserId()
        try await supabase
            .from("flipdesk_ebay_listings")
            .update(Update(
                match_status: "matched",
                matched_item_id: matchedItemId
            ))
            .eq("id", value: orphanId)
            .eq("user_id", value: userId)
            .execute()
    }

    /// Resolves the signed-in user's id for tenant scoping (US-1226). Uses the
    /// same `auth.session` accessor as the other in-service flows
    /// (`TemplateService`, `ConsignorService`); throws if there's no session.
    private func currentUserId() async throws -> String {
        try await supabase.auth.session.user.id.uuidString
    }
}

private extension String {
    var nonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

// US-1240: split a collection into fixed-size chunks for bounded-concurrency
// processing (mirrors the US-1166 helper in DraftsLibraryStore). private so it
// can't collide with a future shared helper.
private extension Array {
    func chunked(into size: Int) -> [[Element]] {
        guard size > 0 else { return [self] }
        return stride(from: 0, to: count, by: size).map {
            Array(self[$0 ..< Swift.min($0 + size, count)])
        }
    }
}

File: android/app/src/main/java/com/gradethread/app/marketplaces/reconciliation/ReconciliationService.kt
package com.gradethread.app.marketplaces.reconciliation

import com.gradethread.app.ui.UiMessage

import com.gradethread.app.R

import androidx.annotation.StringRes

import com.gradethread.app.platform.workspace.WorkspaceScope
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.time.Instant
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * US-1356: matching orphan eBay listings to inventory.
 *
 * All Postgres, no edge: every write here is a plain row change RLS already
 * scopes to the owner. Reads and writes still name the tenant explicitly where
 * the table carries a `user_id`, matching the US-268 convention — a forged row
 * id alone must never flip another seller's listing.
 */
@Singleton
class ReconciliationService @Inject constructor(private val client: SupabaseClient) {

    companion object {
        private const val ORPHANS = "flipdesk_ebay_listings"
        private const val ITEMS = "inventory_items"
        private const val LISTINGS = "listings"

        private const val ORPHAN_COLUMNS =
            "id, ebay_item_id, custom_label, title, current_price, available_quantity, " +
                "listing_url, listing_format, imported_at"

        /**
         * Bounded fan-out for create-all. Each create is an independent
         * three-row write with no shared state, so a small overlap turns a
         * serial march into concurrent round-trips without flooding the pool.
         */
        const val BULK_CONCURRENCY = 4

        /** Ceiling for the banner's count read. */
        const val COUNT_CAP = 200

        val DUPLICATE_LISTING_MESSAGE =
            UiMessage(R.string.reconcile_error_duplicate_listing)
    }

    @Serializable
    private data class IdRow(val id: String)

    @Serializable
    private data class ActiveListingRow(val platform_listing_id: String? = null)

    private fun ownerId(): String? = client.auth.currentUserOrNull()?.id?.let { WorkspaceScope.tenantOwnerId(it) }

    /**
     * How many listings are waiting.
     *
     * Reads ID COLUMNS ONLY, bounded to [COUNT_CAP] + 1 rows. The banner needs a
     * number, not the data, and a seller with three thousand orphans should not
     * pull three thousand rows on every foreground to render one line of text.
     * Past the cap the banner says "200+", which is true — inventing an exact
     * total we didn't measure would not be.
     */
    suspend fun countOrphans(): OrphanCount {
        val owner = ownerId() ?: return OrphanCount()
        val rows = client.from(ORPHANS).select(Columns.raw("id")) {
            filter {
                eq("user_id", owner)
                eq("match_status", "unmatched")
            }
            limit((COUNT_CAP + 1).toLong())
        }.decodeList<IdRow>()
        return OrphanCount(
            value = minOf(rows.size, COUNT_CAP),
            atLeast = rows.size > COUNT_CAP,
        )
    }

    /** Every unmatched listing, newest import first. */
    suspend fun fetchOrphans(): List<OrphanEbayListing> {
        val owner = ownerId() ?: return emptyList()
        return client.from(ORPHANS).select(Columns.raw(ORPHAN_COLUMNS)) {
            filter {
                eq("user_id", owner)
                eq("match_status", "unmatched")
            }
            order("imported_at", Order.DESCENDING)
        }.decodeList()
    }

    /**
     * Mint an inventory item from the orphan and mark it matched.
     *
     * The id is generated CLIENT-side so the write is an idempotent upsert: a
     * retry after a lost response updates the same row instead of minting a
     * second item for one eBay listing.
     */
    suspend fun createItem(
        orphan: OrphanEbayListing,
        title: String? = null,
        sku: String? = null,
        targetPrice: Double? = null,
    ): ReconcileOutcome {
        val owner = ownerId()
            ?: return ReconcileOutcome.Failed(
                orphan.id,
                UiMessage(R.string.reconcile_error_signed_out),
            )
        val itemId = UUID.randomUUID().toString().lowercase()

        return runCatching {
            val fields = mutableMapOf(
                "id" to JsonPrimitive(itemId),
                "user_id" to JsonPrimitive(owner),
                "title" to JsonPrimitive(
                    title?.trim()?.takeIf { it.isNotEmpty() } ?: orphan.suggestedTitle,
                ),
                // eBay says this is live, so the item starts in `listed`. Leaving
                // it in a pre-list stage would offer a publish action that
                // double-lists it.
                "status" to JsonPrimitive("listed"),
            )
            val resolvedSku = sku?.trim()?.takeIf { it.isNotEmpty() }
                ?: orphan.customLabel?.trim()?.takeIf { it.isNotEmpty() }
            fields["sku"] = resolvedSku?.let { JsonPrimitive(it) } ?: JsonNull
            (targetPrice ?: orphan.currentPrice)?.let {
                fields["target_price"] = JsonPrimitive(it)
            }

            client.from(ITEMS).upsert(JsonObject(fields))
            // Best-effort mirror: a failure here must not strand a created item,
            // and the next eBay sync reconciles the listing row anyway.
            runCatching { mirrorListing(itemId, orphan) }
            markMatched(orphan.id, itemId, owner)
            ReconcileOutcome.Created(orphan.id, itemId)
        }.getOrElse { failed(orphan.id, it, R.string.reconcile_error_create) }
    }

    /**
     * Attach the orphan to an item the seller already has.
     *
     * Refuses when that item already carries a DIFFERENT active eBay listing:
     * that is how duplicate active rows are born, and every later bulk action
     * then has two rows to choose between. Re-linking the same listing is fine —
     * the mirror is idempotent.
     */
    suspend fun link(orphan: OrphanEbayListing, itemId: String): ReconcileOutcome {
        val owner = ownerId()
            ?: return ReconcileOutcome.Failed(
                orphan.id,
                UiMessage(R.string.reconcile_error_signed_out),
            )
        return runCatching {
            val active = client.from(LISTINGS)
                .select(Columns.raw("platform_listing_id")) {
                    filter {
                        eq("inventory_item_id", itemId)
                        eq("platform", "ebay")
                        eq("is_active", true)
                    }
                }.decodeList<ActiveListingRow>()

            if (active.any { it.platform_listing_id != null && it.platform_listing_id != orphan.ebayItemId }) {
                return ReconcileOutcome.Failed(orphan.id, DUPLICATE_LISTING_MESSAGE)
            }

            mirrorListing(itemId, orphan)
            markMatched(orphan.id, itemId, owner)
            // Same best-effort reasoning as the mirror on create.
            runCatching { markItemListed(itemId, owner) }
            ReconcileOutcome.Linked(orphan.id, itemId)
        }.getOrElse { failed(orphan.id, it, R.string.reconcile_error_link) }
    }

    /** Drop it from the queue. A re-sync can bring it back. */
    suspend fun ignore(orphan: OrphanEbayListing): ReconcileOutcome {
        val owner = ownerId()
            ?: return ReconcileOutcome.Failed(
                orphan.id,
                UiMessage(R.string.reconcile_error_signed_out),
            )
        return runCatching {
            client.from(ORPHANS).update(
                JsonObject(mapOf("match_status" to JsonPrimitive("ignored"))),
            ) {
                filter {
                    eq("id", orphan.id)
                    eq("user_id", owner)
                }
            }
            ReconcileOutcome.Ignored(orphan.id)
        }.getOrElse { failed(orphan.id, it, R.string.reconcile_error_ignore) }
    }

    /**
     * Create an item for every orphan, a few at a time.
     *
     * One bad row never stops the batch — the failures come back attached to
     * their orphans so the queue can show which ones still need a decision.
     * [onProgress] fires after each chunk so a long run isn't one opaque spinner.
     */
    suspend fun createAll(
        orphans: List<OrphanEbayListing>,
        onProgress: (done: Int, total: Int) -> Unit = { _, _ -> },
    ): ReconcileBulkResult = coroutineScope {
        val outcomes = mutableListOf<ReconcileOutcome>()
        for (chunk in orphans.chunked(BULK_CONCURRENCY)) {
            outcomes += chunk.map { async { createItem(it) } }.awaitAll()
            onProgress(outcomes.size, orphans.size)
        }
        ReconcileBulkResult.from(outcomes)
    }

    // ── Internals ────────────────────────────────────────────────────────────

    /**
     * Mirror the live eBay listing into `listings` so the item's canvas shows
     * where it is listed. Idempotent: updates the existing row for this
     * item + eBay id, else inserts one.
     *
     * `listed_at` is written only on INSERT — an update must not move the date
     * the listing actually went live.
     */
    private suspend fun mirrorListing(itemId: String, orphan: OrphanEbayListing) {
        val existing = client.from(LISTINGS).select(Columns.raw("id")) {
            filter {
                eq("inventory_item_id", itemId)
                eq("platform", "ebay")
                eq("platform_listing_id", orphan.ebayItemId)
            }
            limit(1)
        }.decodeList<IdRow>().firstOrNull()

        val shared = mutableMapOf(
            "platform" to JsonPrimitive("ebay"),
            // The listing was authored on eBay, so eBay owns its editable
            // fields from here (vault/20-domain/sync-source-of-truth.md).
            "listing_origin" to JsonPrimitive("ebay"),
            "listing_price" to JsonPrimitive(orphan.currentPrice ?: 0.0),
            "listing_status" to JsonPrimitive("active"),
            "is_active" to JsonPrimitive(true),
        )
        shared["listing_url"] = orphan.listingUrl?.let { JsonPrimitive(it) } ?: JsonNull
        shared["listing_title"] = orphan.title?.let { JsonPrimitive(it) } ?: JsonNull

        if (existing != null) {
            client.from(LISTINGS).update(JsonObject(shared)) {
                filter { eq("id", existing.id) }
            }
        } else {
            shared["inventory_item_id"] = JsonPrimitive(itemId)
            shared["platform_listing_id"] = JsonPrimitive(orphan.ebayItemId)
            shared["listed_at"] = JsonPrimitive(Instant.now().toString())
            client.from(LISTINGS).insert(JsonObject(shared))
        }
    }

    private suspend fun markMatched(orphanId: String, itemId: String, owner: String) {
        client.from(ORPHANS).update(
            JsonObject(
                mapOf(
                    "match_status" to JsonPrimitive("matched"),
                    "matched_item_id" to JsonPrimitive(itemId),
                ),
            ),
        ) {
            filter {
                eq("id", orphanId)
                eq("user_id", owner)
            }
        }
    }

    private suspend fun markItemListed(itemId: String, owner: String) {
        client.from(ITEMS).update(JsonObject(mapOf("status" to JsonPrimitive("listed")))) {
            filter {
                eq("id", itemId)
                eq("user_id", owner)
            }
        }
    }
}

/**
 * A reconcile failure: the thrown message when there is one, ours otherwise.
 *
 * US-2976: `it.message` is the exception's own words - a PostgREST error, a
 * socket timeout - and cannot be translated here, so it rides as `detail`.
 * Ours is the fallback shown when the throwable said nothing.
 */
private fun failed(orphanId: String, error: Throwable, @StringRes fallback: Int) =
    ReconcileOutcome.Failed(orphanId, UiMessage(fallback, detail = error.message))
